HTTP JSON transcoding doesn't work with recursive types

Using `google.protobuf.Struct` in a proto file and enabling transcoding via `.enableHttpJsonTranscoding(true)` prevents the server from starting. I believe HttpJsonTranscodingService#buildFields is being called infinitely overflowing the call stack.

```
Caused by: java.lang.StackOverflowError: null
	at java.base/java.lang.StringBuilder.<init>(StringBuilder.java:102) ~[na:na]
	at com.linecorp.armeria.server.grpc.HttpJsonTranscodingService.lambda$buildFields$4(HttpJsonTranscodingService.java:291) ~[armeria-grpc-1.13.4.jar:na]
	at java.base/java.util.Arrays$ArrayList.forEach(Arrays.java:4204) ~[na:na]
	at java.base/java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1092) ~[na:na]
	at com.linecorp.armeria.server.grpc.HttpJsonTranscodingService.buildFields(HttpJsonTranscodingService.java:279) ~[armeria-grpc-1.13.4.jar:na]
	at com.linecorp.armeria.server.grpc.HttpJsonTranscodingService.lambda$buildFields$4(HttpJsonTranscodingService.java:323) ~[armeria-grpc-1.13.4.jar:na]
	at java.base/java.util.Arrays$ArrayList.forEach(Arrays.java:4204) ~[na:na]
	at java.base/java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1092) ~[na:na]
	at com.linecorp.armeria.server.grpc.HttpJsonTranscodingService.buildFields(HttpJsonTranscodingService.java:279) ~[armeria-grpc-1.13.4.jar:na]
	at com.linecorp.armeria.server.grpc.HttpJsonTranscodingService.lambda$buildFields$4(HttpJsonTranscodingService.java:323) ~[armeria-grpc-1.13.4.jar:na]
	at java.base/java.util.Arrays$ArrayList.forEach(Arrays.java:4204) ~[na:na]
	at java.base/java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1092) ~[na:na]
	at com.linecorp.armeria.server.grpc.HttpJsonTranscodingService.buildFields(HttpJsonTranscodingService.java:279) ~[armeria-grpc-1.13.4.jar:na]
	at com.linecorp.armeria.server.grpc.HttpJsonTranscodingService.lambda$buildFields$4(HttpJsonTranscodingService.java:323) ~[armeria-grpc-1.13.4.jar:na]
	at java.base/java.util.Arrays$ArrayList.forEach(Arrays.java:4204) ~[na:na]
	at java.base/java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1092) ~[na:na]
	at com.linecorp.armeria.server.grpc.HttpJsonTranscodingService.buildFields(HttpJsonTranscodingService.java:279) ~[armeria-grpc-1.13.4.jar:na]
	at com.linecorp.armeria.server.grpc.HttpJsonTranscodingService.lambda$buildFields$4(HttpJsonTranscodingService.java:323) ~[armeria-grpc-1.13.4.jar:na]
	at java.base/java.util.Arrays$ArrayList.forEach(Arrays.java:4204) ~[na:na]
....
```

```proto
syntax = "proto3";

import "google/api/annotations.proto";
import "google/protobuf/descriptor.proto";
import "google/protobuf/empty.proto";
import "google/protobuf/struct.proto";

message Request {
  google.protobuf.Struct payload = 1;
}

message Response {
  google.protobuf.Struct payload = 2;
}

service SayHelloService {
  rpc SayHello (Request) returns (Response) {
    option (google.api.http) = {get:"/hello"};
  }
}
```

--- a/grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java
+++ b/grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java
@@ -902,5 +902,10 @@ final class HttpJsonTranscodingService extends AbstractUnframedGrpcService
         Descriptor recursiveTypeDescriptor() {
             return recursiveTypeDescriptor;
         }
+
+        @Override
+        public Throwable fillInStackTrace() {
+            return this;
+        }
     }
 }

--- a/grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java
+++ b/grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java
@@ -161,8 +161,8 @@ final class HttpJsonTranscodingService extends AbstractUnframedGrpcService
 
                 final Route route = routeAndVariables.getKey();
                 final List<PathVariable> pathVariables = routeAndVariables.getValue();
-
-                final Map<String, Field> fields = buildFields(methodDesc.getInputType(), ImmutableList.of());
+                final Map<String, Field> fields =
+                        buildFields(methodDesc.getInputType(), ImmutableList.of(), ImmutableSet.of());
 
                 int order = 0;
                 builder.put(route, new TranscodingSpec(order++, httpRule, methodDefinition,
@@ -274,7 +274,9 @@ final class HttpJsonTranscodingService extends AbstractUnframedGrpcService
         return new SimpleImmutableEntry<>(builder.build(), PathVariable.from(segments, pathMappingType));
     }
 
-    private static Map<String, Field> buildFields(Descriptor desc, List<String> parentNames) {
+    private static Map<String, Field> buildFields(Descriptor desc,
+                                                  List<String> parentNames,
+                                                  Set<Descriptor> visitingTypes) {
         final StringJoiner namePrefixJoiner = new StringJoiner(".");
         parentNames.forEach(namePrefixJoiner::add);
         final String namePrefix = namePrefixJoiner.length() == 0 ? "" : namePrefixJoiner.toString() + '.';
@@ -304,6 +306,11 @@ final class HttpJsonTranscodingService extends AbstractUnframedGrpcService
                         break;
                     }
 
+                    if (visitingTypes.contains(field.getMessageType())) {
+                        // Found recursion. No more analysis for this type.
+                        throw new RecursiveTypeException(field.getMessageType());
+                    }
+
                     @Nullable
                     Descriptor typeDesc =
                             desc.getNestedTypes().stream()
@@ -324,10 +331,25 @@ final class HttpJsonTranscodingService extends AbstractUnframedGrpcService
                     checkState(typeDesc != null,
                                "Descriptor for the type '%s' does not exist.",
                                field.getMessageType().getFullName());
-                    builder.putAll(buildFields(typeDesc, ImmutableList.<String>builder()
-                                                                      .addAll(parentNames)
-                                                                      .add(field.getName())
-                                                                      .build()));
+                    try {
+                        builder.putAll(buildFields(typeDesc,
+                                                   ImmutableList.<String>builder()
+                                                                .addAll(parentNames)
+                                                                .add(field.getName())
+                                                                .build(),
+                                                   ImmutableSet.<Descriptor>builder()
+                                                               .addAll(visitingTypes)
+                                                               .add(field.getMessageType())
+                                                               .build()));
+                    } catch (RecursiveTypeException e) {
+                        if (e.recursiveTypeDescriptor() != field.getMessageType()) {
+                            // Re-throw the exception if it is not caused by my field.
+                            throw e;
+                        }
+
+                        builder.put(namePrefix + field.getName(),
+                                    new Field(field, parentNames, JavaType.MESSAGE));
+                    }
                     break;
             }
         });
@@ -336,6 +358,11 @@ final class HttpJsonTranscodingService extends AbstractUnframedGrpcService
 
     @Nullable
     private static JavaType getJavaTypeForWellKnownTypes(FieldDescriptor fd) {
+        // MapField can be sent only via HTTP body.
+        if (fd.isMapField()) {
+            return JavaType.MESSAGE;
+        }
+
         final Descriptor messageType = fd.getMessageType();
         final String fullName = messageType.getFullName();
 
@@ -856,6 +883,23 @@ final class HttpJsonTranscodingService extends AbstractUnframedGrpcService
                  */
                 REFERENCE
             }
+        }
+    }
+
+    /**
+     * Notifies that a recursively nesting type exists.
+     */
+    static class RecursiveTypeException extends IllegalArgumentException {
+        private static final long serialVersionUID = -6764357154559606786L;
+
+        private final Descriptor recursiveTypeDescriptor;
+
+        RecursiveTypeException(Descriptor recursiveTypeDescriptor) {
+            this.recursiveTypeDescriptor = recursiveTypeDescriptor;
+        }
+
+        Descriptor recursiveTypeDescriptor() {
+            return recursiveTypeDescriptor;
         }
     }
 }

--- a/grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java
+++ b/grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java
@@ -276,7 +276,7 @@ final class HttpJsonTranscodingService extends AbstractUnframedGrpcService
 
     private static Map<String, Field> buildFields(Descriptor desc,
                                                   List<String> parentNames,
-                                                  Set<Descriptor> visitingTypes) {
+                                                  Set<Descriptor> visitedTypes) {
         final StringJoiner namePrefixJoiner = new StringJoiner(".");
         parentNames.forEach(namePrefixJoiner::add);
         final String namePrefix = namePrefixJoiner.length() == 0 ? "" : namePrefixJoiner.toString() + '.';
@@ -306,8 +306,9 @@ final class HttpJsonTranscodingService extends AbstractUnframedGrpcService
                         break;
                     }
 
-                    if (visitingTypes.contains(field.getMessageType())) {
+                    if (visitedTypes.contains(field.getMessageType())) {
                         // Found recursion. No more analysis for this type.
+                        // Raise an exception in order to mark the root parameter as JavaType.MESSAGE.
                         throw new RecursiveTypeException(field.getMessageType());
                     }
 
@@ -338,7 +339,7 @@ final class HttpJsonTranscodingService extends AbstractUnframedGrpcService
                                                                 .add(field.getName())
                                                                 .build(),
                                                    ImmutableSet.<Descriptor>builder()
-                                                               .addAll(visitingTypes)
+                                                               .addAll(visitedTypes)
                                                                .add(field.getMessageType())
                                                                .build()));
                     } catch (RecursiveTypeException e) {
@@ -889,7 +890,7 @@ final class HttpJsonTranscodingService extends AbstractUnframedGrpcService
     /**
      * Notifies that a recursively nesting type exists.
      */
-    static class RecursiveTypeException extends IllegalArgumentException {
+    private static class RecursiveTypeException extends IllegalArgumentException {
         private static final long serialVersionUID = -6764357154559606786L;
 
         private final Descriptor recursiveTypeDescriptor;

--- a/grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java
+++ b/grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java
@@ -50,6 +50,7 @@ import com.google.common.base.Strings;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.ImmutableSet;
+import com.google.protobuf.Any;
 import com.google.protobuf.BoolValue;
 import com.google.protobuf.BytesValue;
 import com.google.protobuf.DescriptorProtos.MethodOptions;
@@ -65,10 +66,13 @@ import com.google.protobuf.Duration;
 import com.google.protobuf.FloatValue;
 import com.google.protobuf.Int32Value;
 import com.google.protobuf.Int64Value;
+import com.google.protobuf.ListValue;
 import com.google.protobuf.StringValue;
+import com.google.protobuf.Struct;
 import com.google.protobuf.Timestamp;
 import com.google.protobuf.UInt32Value;
 import com.google.protobuf.UInt64Value;
+import com.google.protobuf.Value;
 
 import com.linecorp.armeria.common.AggregatedHttpRequest;
 import com.linecorp.armeria.common.HttpData;
@@ -354,6 +358,24 @@ final class HttpJsonTranscodingService extends AbstractUnframedGrpcService
             return messageType.getFields().get(0).getJavaType();
         }
 
+        // The messages of the following types can be sent only via HTTP body.
+        if (Struct.getDescriptor().getFullName().equals(fullName) ||
+            ListValue.getDescriptor().getFullName().equals(fullName) ||
+            Value.getDescriptor().getFullName().equals(fullName) ||
+            // google.protobuf.Any message has the following two fields:
+            //   string type_url = 1;
+            //   bytes value = 2;
+            // which look acceptable as HTTP GET parameters, but the client must send the message like below:
+            // {
+            //   "@type": "type.googleapis.com/google.protobuf.Duration",
+            //   "value": "1.212s"
+            // }
+            // There's no specifications about rewriting parameter names, so we will handle
+            // google.protobuf.Any message only when it is sent via HTTP body.
+            Any.getDescriptor().getFullName().equals(fullName)) {
+            return JavaType.MESSAGE;
+        }
+
         return null;
     }
 
@@ -572,6 +594,12 @@ final class HttpJsonTranscodingService extends AbstractUnframedGrpcService
                 // Ignore unknown parameters.
                 continue;
             }
+
+            if (field.javaType == JavaType.MESSAGE) {
+                throw new IllegalArgumentException(
+                        "Unsupported message type: " + field.descriptor.getFullName());
+            }
+
             ObjectNode currentNode = root;
             for (String parentName : field.parentNames) {
                 final JsonNode node = currentNode.get(parentName);

--- a/grpc/src/test/java/com/linecorp/armeria/it/grpc/HttpJsonTranscodingTest.java
+++ b/grpc/src/test/java/com/linecorp/armeria/it/grpc/HttpJsonTranscodingTest.java
@@ -615,6 +615,6 @@ class HttpJsonTranscodingTest {
             throws ExecutionException, InterruptedException {
         final RequestHeaders headers = RequestHeaders.builder().method(HttpMethod.POST).path(path)
                                                      .contentType(MediaType.JSON).build();
-        return webClient.execute(headers, body).aggregate().get();
+        return webClient.execute(headers, body).aggregate().join();
     }
 }

--- a/grpc/src/test/java/com/linecorp/armeria/it/grpc/HttpJsonTranscodingTest.java
+++ b/grpc/src/test/java/com/linecorp/armeria/it/grpc/HttpJsonTranscodingTest.java
@@ -33,7 +33,6 @@ import org.junit.jupiter.api.extension.RegisterExtension;
 
 import com.fasterxml.jackson.core.JsonProcessingException;
 import com.fasterxml.jackson.core.TreeNode;
-import com.fasterxml.jackson.databind.JsonMappingException;
 import com.fasterxml.jackson.databind.JsonNode;
 import com.fasterxml.jackson.databind.ObjectMapper;
 
@@ -406,7 +405,7 @@ class HttpJsonTranscodingTest {
     }
 
     @Test
-    void shouldAcceptWrappers2() throws JsonProcessingException, JsonMappingException {
+    void shouldAcceptWrappers2() throws JsonProcessingException {
         final QueryParamsBuilder query = QueryParams.builder();
         query.add("doubleVal", String.valueOf(123.456d))
              .add("int64Val", String.valueOf(Long.MAX_VALUE))
@@ -428,7 +427,7 @@ class HttpJsonTranscodingTest {
     }
 
     @Test
-    void shouldAcceptStruct() throws JsonProcessingException, JsonMappingException {
+    void shouldAcceptStruct() throws JsonProcessingException {
         final String jsonContent = "{\"intVal\": 1, \"stringVal\": \"1\"}";
         final AggregatedHttpResponse response = jsonPostRequest(webClient, "/v1/echo/struct", jsonContent);
         final JsonNode root = mapper.readTree(response.contentUtf8());
@@ -439,7 +438,7 @@ class HttpJsonTranscodingTest {
     }
 
     @Test
-    void shouldAcceptListValue_String() throws JsonProcessingException, JsonMappingException {
+    void shouldAcceptListValue_String() throws JsonProcessingException {
         final String jsonContent = "[\"1\", \"2\"]";
         final AggregatedHttpResponse response = jsonPostRequest(webClient, "/v1/echo/list_value", jsonContent);
         final JsonNode root = mapper.readTree(response.contentUtf8());
@@ -450,7 +449,7 @@ class HttpJsonTranscodingTest {
     }
 
     @Test
-    void shouldAcceptListValue_Number() throws JsonProcessingException, JsonMappingException {
+    void shouldAcceptListValue_Number() throws JsonProcessingException {
         final String jsonContent = "[1, 2]";
         final AggregatedHttpResponse response = jsonPostRequest(webClient, "/v1/echo/list_value", jsonContent);
         final JsonNode root = mapper.readTree(response.contentUtf8());
@@ -461,7 +460,7 @@ class HttpJsonTranscodingTest {
     }
 
     @Test
-    void shouldAcceptValue_String() throws JsonProcessingException, JsonMappingException {
+    void shouldAcceptValue_String() throws JsonProcessingException {
         final String jsonContent = "\"1\"";
         final AggregatedHttpResponse response = jsonPostRequest(webClient, "/v1/echo/value", jsonContent);
         final JsonNode root = mapper.readTree(response.contentUtf8());
@@ -469,7 +468,7 @@ class HttpJsonTranscodingTest {
     }
 
     @Test
-    void shouldAcceptValue_Number() throws JsonProcessingException, JsonMappingException {
+    void shouldAcceptValue_Number() throws JsonProcessingException {
         final String jsonContent = "1";
         final AggregatedHttpResponse response = jsonPostRequest(webClient, "/v1/echo/value", jsonContent);
         final JsonNode root = mapper.readTree(response.contentUtf8());
@@ -477,7 +476,7 @@ class HttpJsonTranscodingTest {
     }
 
     @Test
-    void shouldAcceptAny() throws JsonProcessingException, JsonMappingException {
+    void shouldAcceptAny() throws JsonProcessingException {
         final String jsonContent =
                 '{' +
                 "  \"@type\": \"type.googleapis.com/google.protobuf.Duration\"," +

--- a/grpc/src/test/java/com/linecorp/armeria/it/grpc/HttpJsonTranscodingTest.java
+++ b/grpc/src/test/java/com/linecorp/armeria/it/grpc/HttpJsonTranscodingTest.java
@@ -25,12 +25,14 @@ import java.time.ZonedDateTime;
 import java.time.format.DateTimeFormatter;
 import java.util.Base64;
 import java.util.Optional;
+import java.util.concurrent.ExecutionException;
 import java.util.stream.Collectors;
 import java.util.stream.StreamSupport;
 
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.RegisterExtension;
 
+import com.fasterxml.jackson.core.TreeNode;
 import com.fasterxml.jackson.databind.JsonNode;
 import com.fasterxml.jackson.databind.ObjectMapper;
 
@@ -47,8 +49,16 @@ import com.linecorp.armeria.common.RequestHeaders;
 import com.linecorp.armeria.common.grpc.GrpcSerializationFormats;
 import com.linecorp.armeria.grpc.testing.HttpJsonTranscodingTestServiceGrpc.HttpJsonTranscodingTestServiceBlockingStub;
 import com.linecorp.armeria.grpc.testing.HttpJsonTranscodingTestServiceGrpc.HttpJsonTranscodingTestServiceImplBase;
+import com.linecorp.armeria.grpc.testing.Transcoding.EchoAnyRequest;
+import com.linecorp.armeria.grpc.testing.Transcoding.EchoAnyResponse;
+import com.linecorp.armeria.grpc.testing.Transcoding.EchoListValueRequest;
+import com.linecorp.armeria.grpc.testing.Transcoding.EchoListValueResponse;
+import com.linecorp.armeria.grpc.testing.Transcoding.EchoStructRequest;
+import com.linecorp.armeria.grpc.testing.Transcoding.EchoStructResponse;
 import com.linecorp.armeria.grpc.testing.Transcoding.EchoTimestampAndDurationRequest;
 import com.linecorp.armeria.grpc.testing.Transcoding.EchoTimestampAndDurationResponse;
+import com.linecorp.armeria.grpc.testing.Transcoding.EchoValueRequest;
+import com.linecorp.armeria.grpc.testing.Transcoding.EchoValueResponse;
 import com.linecorp.armeria.grpc.testing.Transcoding.EchoWrappersRequest;
 import com.linecorp.armeria.grpc.testing.Transcoding.EchoWrappersResponse;
 import com.linecorp.armeria.grpc.testing.Transcoding.GetMessageRequestV1;
@@ -160,6 +170,31 @@ class HttpJsonTranscodingTest {
                 assertThat(request.getBytesVal().getValue().toStringUtf8()).isEqualTo(testBytesValue);
             }
             responseObserver.onNext(builder.build());
+            responseObserver.onCompleted();
+        }
+
+        @Override
+        public void echoStruct(EchoStructRequest request, StreamObserver<EchoStructResponse> responseObserver) {
+            responseObserver.onNext(EchoStructResponse.newBuilder().setValue(request.getValue()).build());
+            responseObserver.onCompleted();
+        }
+
+        @Override
+        public void echoListValue(EchoListValueRequest request,
+                                  StreamObserver<EchoListValueResponse> responseObserver) {
+            responseObserver.onNext(EchoListValueResponse.newBuilder().setValue(request.getValue()).build());
+            responseObserver.onCompleted();
+        }
+
+        @Override
+        public void echoValue(EchoValueRequest request, StreamObserver<EchoValueResponse> responseObserver) {
+            responseObserver.onNext(EchoValueResponse.newBuilder().setValue(request.getValue()).build());
+            responseObserver.onCompleted();
+        }
+
+        @Override
+        public void echoAny(EchoAnyRequest request, StreamObserver<EchoAnyResponse> responseObserver) {
+            responseObserver.onNext(EchoAnyResponse.newBuilder().setValue(request.getValue()).build());
             responseObserver.onCompleted();
         }
     }
@@ -374,6 +409,82 @@ class HttpJsonTranscodingTest {
     }
 
     @Test
+    void shouldAcceptStruct() throws Exception {
+        final String jsonContent = "{\"intVal\": 1, \"stringVal\": \"1\"}";
+        final AggregatedHttpResponse response = jsonPostRequest(webClient, "/v1/echo/struct", jsonContent);
+        final JsonNode root = mapper.readTree(response.contentUtf8());
+        final JsonNode value = root.get("value");
+        assertThat(value).isNotNull().matches(v -> ((TreeNode) v).isObject());
+        assertThat(value.get("intVal").asInt()).isOne();
+        assertThat(value.get("stringVal").asText()).isEqualTo("1");
+    }
+
+    @Test
+    void shouldAcceptListValue_String() throws Exception {
+        final String jsonContent = "[\"1\", \"2\"]";
+        final AggregatedHttpResponse response = jsonPostRequest(webClient, "/v1/echo/list_value", jsonContent);
+        final JsonNode root = mapper.readTree(response.contentUtf8());
+        final JsonNode value = root.get("value");
+        assertThat(value.isArray()).isTrue();
+        assertThat(value.get(0).asText()).isEqualTo("1");
+        assertThat(value.get(1).asText()).isEqualTo("2");
+    }
+
+    @Test
+    void shouldAcceptListValue_Number() throws Exception {
+        final String jsonContent = "[1, 2]";
+        final AggregatedHttpResponse response = jsonPostRequest(webClient, "/v1/echo/list_value", jsonContent);
+        final JsonNode root = mapper.readTree(response.contentUtf8());
+        final JsonNode value = root.get("value");
+        assertThat(value.isArray()).isTrue();
+        assertThat(value.get(0).asInt()).isEqualTo(1);
+        assertThat(value.get(1).asInt()).isEqualTo(2);
+    }
+
+    @Test
+    void shouldAcceptValue_String() throws Exception {
+        final String jsonContent = "\"1\"";
+        final AggregatedHttpResponse response = jsonPostRequest(webClient, "/v1/echo/value", jsonContent);
+        final JsonNode root = mapper.readTree(response.contentUtf8());
+        assertThat(root.get("value").asText()).isEqualTo("1");
+    }
+
+    @Test
+    void shouldAcceptValue_Number() throws Exception {
+        final String jsonContent = "1";
+        final AggregatedHttpResponse response = jsonPostRequest(webClient, "/v1/echo/value", jsonContent);
+        final JsonNode root = mapper.readTree(response.contentUtf8());
+        assertThat(root.get("value").asInt()).isEqualTo(1);
+    }
+
+    @Test
+    void shouldAcceptAny() throws Exception {
+        final String jsonContent =
+                '{' +
+                "  \"@type\": \"type.googleapis.com/google.protobuf.Duration\"," +
+                "  \"value\": \"1.212s\"" +
+                '}';
+        final AggregatedHttpResponse response = jsonPostRequest(webClient, "/v1/echo/any", jsonContent);
+        final JsonNode root = mapper.readTree(response.contentUtf8());
+        final JsonNode value = root.get("value");
+        assertThat(value).isNotNull().matches(v -> ((TreeNode) v).isObject());
+        assertThat(value.get("@type").asText()).isEqualTo("type.googleapis.com/google.protobuf.Duration");
+        assertThat(value.get("value").asText()).isEqualTo("1.212s");
+    }
+
+    @Test
+    void shouldDenyHttpGetParameters_Struct_Value_ListValue_Any() throws Exception {
+        assertThat(webClient.get("/v1/echo/struct?value=1").aggregate().get().status())
+                .isEqualTo(HttpStatus.BAD_REQUEST);
+        assertThat(webClient.get("/v1/echo/list_value?value=1&value=2").aggregate().get().status())
+                .isEqualTo(HttpStatus.BAD_REQUEST);
+        assertThat(webClient.get("/v1/echo/value?value=1").aggregate().get().status())
+                .isEqualTo(HttpStatus.BAD_REQUEST);
+        assertThat(webClient.get("/v1/echo/any?value=1").aggregate().get().status())
+                .isEqualTo(HttpStatus.BAD_REQUEST);
+    }
+
+    @Test
     void shouldAcceptNaNAndInfinity() throws Exception {
         final QueryParamsBuilder query = QueryParams.builder();
         query.add("doubleVal", "NaN")
@@ -433,5 +544,12 @@ class HttpJsonTranscodingTest {
 
     private static String pathMapping(JsonNode method) {
         return method.get("endpoints").get(0).get("pathMapping").asText();
+    }
+
+    private static AggregatedHttpResponse jsonPostRequest(WebClient webClient, String path, String body)
+            throws ExecutionException, InterruptedException {
+        final RequestHeaders headers = RequestHeaders.builder().method(HttpMethod.POST).path(path)
+                                                     .contentType(MediaType.JSON).build();
+        return webClient.execute(headers, body).aggregate().get();
     }
 }

--- a/grpc/src/test/java/com/linecorp/armeria/it/grpc/HttpJsonTranscodingTest.java
+++ b/grpc/src/test/java/com/linecorp/armeria/it/grpc/HttpJsonTranscodingTest.java
@@ -25,14 +25,15 @@ import java.time.ZonedDateTime;
 import java.time.format.DateTimeFormatter;
 import java.util.Base64;
 import java.util.Optional;
-import java.util.concurrent.ExecutionException;
 import java.util.stream.Collectors;
 import java.util.stream.StreamSupport;
 
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.RegisterExtension;
 
+import com.fasterxml.jackson.core.JsonProcessingException;
 import com.fasterxml.jackson.core.TreeNode;
+import com.fasterxml.jackson.databind.JsonMappingException;
 import com.fasterxml.jackson.databind.JsonNode;
 import com.fasterxml.jackson.databind.ObjectMapper;
 
@@ -244,21 +245,21 @@ class HttpJsonTranscodingTest {
     final WebClient webClient = WebClient.builder(server.httpUri()).build();
 
     @Test
-    void shouldGetMessageV1ByGrpcClient() throws Exception {
+    void shouldGetMessageV1ByGrpcClient() {
         final Message message = grpcClient.getMessageV1(
                 GetMessageRequestV1.newBuilder().setName("messages/1").build());
         assertThat(message.getText()).isEqualTo("messages/1");
     }
 
     @Test
-    void shouldGetMessageV1ByWebClient() throws Exception {
-        final AggregatedHttpResponse response = webClient.get("/v1/messages/1").aggregate().get();
+    void shouldGetMessageV1ByWebClient() throws JsonProcessingException {
+        final AggregatedHttpResponse response = webClient.get("/v1/messages/1").aggregate().join();
         final JsonNode root = mapper.readTree(response.contentUtf8());
         assertThat(root.get("text").asText()).isEqualTo("messages/1");
     }
 
     @Test
-    void shouldGetMessageV2ByGrpcClient() throws Exception {
+    void shouldGetMessageV2ByGrpcClient() {
         final Message message = grpcClient.getMessageV2(
                 GetMessageRequestV2.newBuilder()
                                    .setMessageId("1")
@@ -270,26 +271,27 @@ class HttpJsonTranscodingTest {
     }
 
     @Test
-    void shouldGetMessageV2ByWebClient() throws Exception {
+    void shouldGetMessageV2ByWebClient() throws JsonProcessingException {
         final AggregatedHttpResponse response =
                 webClient.get("/v2/messages/1?revision=999&sub.subfield=sub&type=DETAIL")
-                         .aggregate().get();
+                         .aggregate().join();
         final JsonNode root = mapper.readTree(response.contentUtf8());
         assertThat(root.get("text").asText()).isEqualTo("1:999:sub:DETAIL");
     }
 
     @Test
-    void shouldGetMessageV2ByWebClient_GetDefaultValueIfUnknownEnumIsSpecified() throws Exception {
+    void shouldGetMessageV2ByWebClient_GetDefaultValueIfUnknownEnumIsSpecified()
+            throws JsonProcessingException {
         final AggregatedHttpResponse response =
                 webClient.get("/v2/messages/1?revision=999&sub.subfield=sub&type=UNKNOWN")
-                         .aggregate().get();
+                         .aggregate().join();
         final JsonNode root = mapper.readTree(response.contentUtf8());
         // Return a default enum(value 0).
         assertThat(root.get("text").asText()).isEqualTo("1:999:sub:SIMPLE");
     }
 
     @Test
-    void shouldGetMessageV3ByGrpcClient() throws Exception {
+    void shouldGetMessageV3ByGrpcClient() {
         final Message message = grpcClient.getMessageV3(
                 GetMessageRequestV3.newBuilder()
                                    .setMessageId("1")
@@ -299,25 +301,25 @@ class HttpJsonTranscodingTest {
     }
 
     @Test
-    void shouldGetMessageV3ByWebClient() throws Exception {
+    void shouldGetMessageV3ByWebClient() throws JsonProcessingException {
         final AggregatedHttpResponse response =
                 webClient.get("/v3/messages/1?revision=2&revision=3&revision=4")
-                         .aggregate().get();
+                         .aggregate().join();
         final JsonNode root = mapper.readTree(response.contentUtf8());
         assertThat(root.get("text").asText()).isEqualTo("1:2:3:4");
     }
 
     @Test
-    void shouldGetMessageV3ByWebClient_CheckOrder() throws Exception {
+    void shouldGetMessageV3ByWebClient_CheckOrder() throws JsonProcessingException {
         final AggregatedHttpResponse response =
                 webClient.get("/v3/messages/1?revision=4&revision=3&revision=2")
-                         .aggregate().get();
+                         .aggregate().join();
         final JsonNode root = mapper.readTree(response.contentUtf8());
         assertThat(root.get("text").asText()).isEqualTo("1:4:3:2");
     }
 
     @Test
-    void shouldUpdateMessageV1ByGrpcClient() throws Exception {
+    void shouldUpdateMessageV1ByGrpcClient() {
         final Message message = grpcClient.updateMessageV1(
                 UpdateMessageRequestV1.newBuilder()
                                       .setMessageId("1")
@@ -327,7 +329,7 @@ class HttpJsonTranscodingTest {
     }
 
     @Test
-    void shouldUpdateMessageV1ByWebClient() throws Exception {
+    void shouldUpdateMessageV1ByWebClient() throws JsonProcessingException {
         final AggregatedHttpResponse response =
                 webClient.execute(RequestHeaders.builder()
                                                 .method(HttpMethod.PATCH)
@@ -335,20 +337,20 @@ class HttpJsonTranscodingTest {
                                                 .contentType(MediaType.JSON)
                                                 .build(),
                                   HttpData.ofUtf8("{\"text\": \"v1\"}"))
-                         .aggregate().get();
+                         .aggregate().join();
         final JsonNode root = mapper.readTree(response.contentUtf8());
         assertThat(root.get("text").asText()).isEqualTo("1:v1");
     }
 
     @Test
-    void shouldUpdateMessageV2ByGrpcClient() throws Exception {
+    void shouldUpdateMessageV2ByGrpcClient() {
         final Message message = grpcClient.updateMessageV2(Message.newBuilder().setText("v2").build());
         // There's no way to get 'message_id' from a gRPC request.
         assertThat(message.getText()).isEqualTo("no_id:v2");
     }
 
     @Test
-    void shouldUpdateMessageV2ByWebClient() throws Exception {
+    void shouldUpdateMessageV2ByWebClient() throws JsonProcessingException {
         final AggregatedHttpResponse response =
                 webClient.execute(RequestHeaders.builder()
                                                 .method(HttpMethod.PATCH)
@@ -356,25 +358,25 @@ class HttpJsonTranscodingTest {
                                                 .contentType(MediaType.JSON)
                                                 .build(),
                                   HttpData.ofUtf8("{\"text\": \"v2\"}"))
-                         .aggregate().get();
+                         .aggregate().join();
         final JsonNode root = mapper.readTree(response.contentUtf8());
         assertThat(root.get("text").asText()).isEqualTo("1:v2");
     }
 
     @Test
-    void shouldAcceptRfc3339TimeFormatAndDuration() throws Exception {
+    void shouldAcceptRfc3339TimeFormatAndDuration() throws JsonProcessingException {
         final String timestamp = ZonedDateTime.now().format(DateTimeFormatter.ISO_INSTANT);
         final String duration = "1.000340012s";
 
         final AggregatedHttpResponse response =
-                webClient.get("/v1/echo/" + timestamp + '/' + duration).aggregate().get();
+                webClient.get("/v1/echo/" + timestamp + '/' + duration).aggregate().join();
         final JsonNode root = mapper.readTree(response.contentUtf8());
         assertThat(root.get("timestamp").asText()).isEqualTo(timestamp);
         assertThat(root.get("duration").asText()).isEqualTo(duration);
     }
 
     @Test
-    void shouldAcceptWrappers1() throws Exception {
+    void shouldAcceptWrappers1() throws JsonProcessingException {
         final String bytesValue = new String(
                 Base64.getEncoder().encode(testBytesValue.getBytes(StandardCharsets.UTF_8)));
 
@@ -390,7 +392,7 @@ class HttpJsonTranscodingTest {
              .add("bytesVal", bytesValue);
 
         final AggregatedHttpResponse response =
-                webClient.get("/v1/echo/wrappers?" + query.toQueryString()).aggregate().get();
+                webClient.get("/v1/echo/wrappers?" + query.toQueryString()).aggregate().join();
         final JsonNode root = mapper.readTree(response.contentUtf8());
         assertThat(root.get("doubleVal").asDouble()).isEqualTo(123.456d, withPrecision(0.001d));
         assertThat(root.get("floatVal").asDouble()).isEqualTo(123.456f, withPrecision(0.001d));
@@ -404,7 +406,7 @@ class HttpJsonTranscodingTest {
     }
 
     @Test
-    void shouldAcceptWrappers2() throws Exception {
+    void shouldAcceptWrappers2() throws JsonProcessingException, JsonMappingException {
         final QueryParamsBuilder query = QueryParams.builder();
         query.add("doubleVal", String.valueOf(123.456d))
              .add("int64Val", String.valueOf(Long.MAX_VALUE))
@@ -412,7 +414,7 @@ class HttpJsonTranscodingTest {
              .add("stringVal", "string");
 
         final AggregatedHttpResponse response =
-                webClient.get("/v1/echo/wrappers?" + query.toQueryString()).aggregate().get();
+                webClient.get("/v1/echo/wrappers?" + query.toQueryString()).aggregate().join();
         final JsonNode root = mapper.readTree(response.contentUtf8());
         assertThat(root.get("doubleVal").asDouble()).isEqualTo(123.456d, withPrecision(0.001d));
         assertThat(root.get("floatVal")).isNull();
@@ -426,7 +428,7 @@ class HttpJsonTranscodingTest {
     }
 
     @Test
-    void shouldAcceptStruct() throws Exception {
+    void shouldAcceptStruct() throws JsonProcessingException, JsonMappingException {
         final String jsonContent = "{\"intVal\": 1, \"stringVal\": \"1\"}";
         final AggregatedHttpResponse response = jsonPostRequest(webClient, "/v1/echo/struct", jsonContent);
         final JsonNode root = mapper.readTree(response.contentUtf8());
@@ -437,7 +439,7 @@ class HttpJsonTranscodingTest {
     }
 
     @Test
-    void shouldAcceptListValue_String() throws Exception {
+    void shouldAcceptListValue_String() throws JsonProcessingException, JsonMappingException {
         final String jsonContent = "[\"1\", \"2\"]";
         final AggregatedHttpResponse response = jsonPostRequest(webClient, "/v1/echo/list_value", jsonContent);
         final JsonNode root = mapper.readTree(response.contentUtf8());
@@ -448,7 +450,7 @@ class HttpJsonTranscodingTest {
     }
 
     @Test
-    void shouldAcceptListValue_Number() throws Exception {
+    void shouldAcceptListValue_Number() throws JsonProcessingException, JsonMappingException {
         final String jsonContent = "[1, 2]";
         final AggregatedHttpResponse response = jsonPostRequest(webClient, "/v1/echo/list_value", jsonContent);
         final JsonNode root = mapper.readTree(response.contentUtf8());
@@ -459,7 +461,7 @@ class HttpJsonTranscodingTest {
     }
 
     @Test
-    void shouldAcceptValue_String() throws Exception {
+    void shouldAcceptValue_String() throws JsonProcessingException, JsonMappingException {
         final String jsonContent = "\"1\"";
         final AggregatedHttpResponse response = jsonPostRequest(webClient, "/v1/echo/value", jsonContent);
         final JsonNode root = mapper.readTree(response.contentUtf8());
@@ -467,7 +469,7 @@ class HttpJsonTranscodingTest {
     }
 
     @Test
-    void shouldAcceptValue_Number() throws Exception {
+    void shouldAcceptValue_Number() throws JsonProcessingException, JsonMappingException {
         final String jsonContent = "1";
         final AggregatedHttpResponse response = jsonPostRequest(webClient, "/v1/echo/value", jsonContent);
         final JsonNode root = mapper.readTree(response.contentUtf8());
@@ -475,7 +477,7 @@ class HttpJsonTranscodingTest {
     }
 
     @Test
-    void shouldAcceptAny() throws Exception {
+    void shouldAcceptAny() throws JsonProcessingException, JsonMappingException {
         final String jsonContent =
                 '{' +
                 "  \"@type\": \"type.googleapis.com/google.protobuf.Duration\"," +
@@ -490,19 +492,19 @@ class HttpJsonTranscodingTest {
     }
 
     @Test
-    void shouldDenyHttpGetParameters_Struct_Value_ListValue_Any() throws Exception {
-        assertThat(webClient.get("/v1/echo/struct?value=1").aggregate().get().status())
+    void shouldDenyHttpGetParameters_Struct_Value_ListValue_Any() {
+        assertThat(webClient.get("/v1/echo/struct?value=1").aggregate().join().status())
                 .isEqualTo(HttpStatus.BAD_REQUEST);
-        assertThat(webClient.get("/v1/echo/list_value?value=1&value=2").aggregate().get().status())
+        assertThat(webClient.get("/v1/echo/list_value?value=1&value=2").aggregate().join().status())
                 .isEqualTo(HttpStatus.BAD_REQUEST);
-        assertThat(webClient.get("/v1/echo/value?value=1").aggregate().get().status())
+        assertThat(webClient.get("/v1/echo/value?value=1").aggregate().join().status())
                 .isEqualTo(HttpStatus.BAD_REQUEST);
-        assertThat(webClient.get("/v1/echo/any?value=1").aggregate().get().status())
+        assertThat(webClient.get("/v1/echo/any?value=1").aggregate().join().status())
                 .isEqualTo(HttpStatus.BAD_REQUEST);
     }
 
     @Test
-    void shouldAcceptRecursive() throws Exception {
+    void shouldAcceptRecursive() throws JsonProcessingException {
         final String jsonContent =
                 '{' +
                 "  \"value\": \"a\"," +
@@ -521,7 +523,7 @@ class HttpJsonTranscodingTest {
     }
 
     @Test
-    void shouldAcceptRecursive2() throws Exception {
+    void shouldAcceptRecursive2() throws JsonProcessingException {
         final String jsonContent =
                 '{' +
                 "  \"value\": \"a\"," +
@@ -538,40 +540,40 @@ class HttpJsonTranscodingTest {
     }
 
     @Test
-    void shoudDenyRecursiveViaHttpGet() throws Exception {
-        assertThat(webClient.get("/v1/echo/recursive?value=a").aggregate().get().status())
+    void shoudDenyRecursiveViaHttpGet() {
+        assertThat(webClient.get("/v1/echo/recursive?value=a").aggregate().join().status())
                 .isEqualTo(HttpStatus.BAD_REQUEST);
-        assertThat(webClient.get("/v1/echo/recursive2?nested=a").aggregate().get().status())
+        assertThat(webClient.get("/v1/echo/recursive2?nested=a").aggregate().join().status())
                 .isEqualTo(HttpStatus.BAD_REQUEST);
 
         // Note that the parameter will be ignored because it won't be matched with any fields.
-        assertThat(webClient.get("/v1/echo/recursive?value.nested.value=a").aggregate().get().status())
+        assertThat(webClient.get("/v1/echo/recursive?value.nested.value=a").aggregate().join().status())
                 .isEqualTo(HttpStatus.OK);
     }
 
     @Test
-    void shouldAcceptNaNAndInfinity() throws Exception {
+    void shouldAcceptNaNAndInfinity() throws JsonProcessingException {
         final QueryParamsBuilder query = QueryParams.builder();
         query.add("doubleVal", "NaN")
              .add("floatVal", "Infinity");
 
         final AggregatedHttpResponse response =
-                webClient.get("/v1/echo/wrappers?" + query.toQueryString()).aggregate().get();
+                webClient.get("/v1/echo/wrappers?" + query.toQueryString()).aggregate().join();
         final JsonNode root = mapper.readTree(response.contentUtf8());
         assertThat(root.get("doubleVal").asDouble()).isNaN();
         assertThat(root.get("floatVal").asDouble()).isInfinite();
     }
 
     @Test
-    void shouldReturnMethodNotAllowed() throws Exception {
-        final AggregatedHttpResponse response = webClient.get("/foo/").aggregate().get();
+    void shouldReturnMethodNotAllowed() {
+        final AggregatedHttpResponse response = webClient.get("/foo/").aggregate().join();
         // Because the FramedGrpcService only support HTTP POST.
         assertThat(response.status()).isEqualTo(HttpStatus.METHOD_NOT_ALLOWED);
     }
 
     @Test
-    void shouldBeIntegratedWithDocService() throws Exception {
-        final AggregatedHttpResponse response = webClient.get("/docs/specification.json").aggregate().get();
+    void shouldBeIntegratedWithDocService() throws JsonProcessingException {
+        final AggregatedHttpResponse response = webClient.get("/docs/specification.json").aggregate().join();
         final JsonNode root = mapper.readTree(response.contentUtf8());
         final JsonNode methods =
                 StreamSupport.stream(root.get("services").spliterator(), false)
@@ -611,8 +613,7 @@ class HttpJsonTranscodingTest {
         return method.get("endpoints").get(0).get("pathMapping").asText();
     }
 
-    private static AggregatedHttpResponse jsonPostRequest(WebClient webClient, String path, String body)
-            throws ExecutionException, InterruptedException {
+    private static AggregatedHttpResponse jsonPostRequest(WebClient webClient, String path, String body) {
         final RequestHeaders headers = RequestHeaders.builder().method(HttpMethod.POST).path(path)
                                                      .contentType(MediaType.JSON).build();
         return webClient.execute(headers, body).aggregate().join();

--- a/grpc/src/test/java/com/linecorp/armeria/it/grpc/HttpJsonTranscodingTest.java
+++ b/grpc/src/test/java/com/linecorp/armeria/it/grpc/HttpJsonTranscodingTest.java
@@ -53,6 +53,8 @@ import com.linecorp.armeria.grpc.testing.Transcoding.EchoAnyRequest;
 import com.linecorp.armeria.grpc.testing.Transcoding.EchoAnyResponse;
 import com.linecorp.armeria.grpc.testing.Transcoding.EchoListValueRequest;
 import com.linecorp.armeria.grpc.testing.Transcoding.EchoListValueResponse;
+import com.linecorp.armeria.grpc.testing.Transcoding.EchoRecursiveRequest;
+import com.linecorp.armeria.grpc.testing.Transcoding.EchoRecursiveResponse;
 import com.linecorp.armeria.grpc.testing.Transcoding.EchoStructRequest;
 import com.linecorp.armeria.grpc.testing.Transcoding.EchoStructResponse;
 import com.linecorp.armeria.grpc.testing.Transcoding.EchoTimestampAndDurationRequest;
@@ -195,6 +197,13 @@ class HttpJsonTranscodingTest {
         @Override
         public void echoAny(EchoAnyRequest request, StreamObserver<EchoAnyResponse> responseObserver) {
             responseObserver.onNext(EchoAnyResponse.newBuilder().setValue(request.getValue()).build());
+            responseObserver.onCompleted();
+        }
+
+        @Override
+        public void echoRecursive(EchoRecursiveRequest request,
+                                  StreamObserver<EchoRecursiveResponse> responseObserver) {
+            responseObserver.onNext(EchoRecursiveResponse.newBuilder().setValue(request.getValue()).build());
             responseObserver.onCompleted();
         }
     }
@@ -482,6 +491,35 @@ class HttpJsonTranscodingTest {
                 .isEqualTo(HttpStatus.BAD_REQUEST);
         assertThat(webClient.get("/v1/echo/any?value=1").aggregate().get().status())
                 .isEqualTo(HttpStatus.BAD_REQUEST);
+    }
+
+    @Test
+    void shouldAcceptRecursive() throws Exception {
+        final String jsonContent =
+                '{' +
+                "  \"value\": \"a\"," +
+                "  \"nested\": {" +
+                "    \"value\": \"b\"" +
+                "  }" +
+                '}';
+        final AggregatedHttpResponse response = jsonPostRequest(webClient, "/v1/echo/recursive", jsonContent);
+        final JsonNode root = mapper.readTree(response.contentUtf8());
+        final JsonNode value = root.get("value");
+        assertThat(value).isNotNull().matches(v -> ((TreeNode) v).isObject());
+        assertThat(value.get("value").asText()).isEqualTo("a");
+        final JsonNode nested = value.get("nested");
+        assertThat(nested).isNotNull().matches(v -> ((TreeNode) v).isObject());
+        assertThat(nested.get("value").asText()).isEqualTo("b");
+    }
+
+    @Test
+    void shoudDenyRecursiveViaHttpGet() throws Exception {
+        assertThat(webClient.get("/v1/echo/recursive?value=a").aggregate().get().status())
+                .isEqualTo(HttpStatus.BAD_REQUEST);
+
+        // Note that the parameter will be ignored because it won't be matched with any fields.
+        assertThat(webClient.get("/v1/echo/recursive?value.nested.value=a").aggregate().get().status())
+                .isEqualTo(HttpStatus.OK);
     }
 
     @Test

--- a/grpc/src/test/java/com/linecorp/armeria/it/grpc/HttpJsonTranscodingTest.java
+++ b/grpc/src/test/java/com/linecorp/armeria/it/grpc/HttpJsonTranscodingTest.java
@@ -69,6 +69,7 @@ import com.linecorp.armeria.grpc.testing.Transcoding.GetMessageRequestV2.SubMess
 import com.linecorp.armeria.grpc.testing.Transcoding.GetMessageRequestV3;
 import com.linecorp.armeria.grpc.testing.Transcoding.Message;
 import com.linecorp.armeria.grpc.testing.Transcoding.MessageType;
+import com.linecorp.armeria.grpc.testing.Transcoding.Recursive;
 import com.linecorp.armeria.grpc.testing.Transcoding.UpdateMessageRequestV1;
 import com.linecorp.armeria.internal.common.JacksonUtil;
 import com.linecorp.armeria.internal.server.grpc.GrpcDocServicePlugin;
@@ -204,6 +205,13 @@ class HttpJsonTranscodingTest {
         public void echoRecursive(EchoRecursiveRequest request,
                                   StreamObserver<EchoRecursiveResponse> responseObserver) {
             responseObserver.onNext(EchoRecursiveResponse.newBuilder().setValue(request.getValue()).build());
+            responseObserver.onCompleted();
+        }
+
+        @Override
+        public void echoRecursive2(Recursive request,
+                                   StreamObserver<Recursive> responseObserver) {
+            responseObserver.onNext(request);
             responseObserver.onCompleted();
         }
     }
@@ -513,8 +521,27 @@ class HttpJsonTranscodingTest {
     }
 
     @Test
+    void shouldAcceptRecursive2() throws Exception {
+        final String jsonContent =
+                '{' +
+                "  \"value\": \"a\"," +
+                "  \"nested\": {" +
+                "    \"value\": \"b\"" +
+                "  }" +
+                '}';
+        final AggregatedHttpResponse response = jsonPostRequest(webClient, "/v1/echo/recursive2", jsonContent);
+        final JsonNode root = mapper.readTree(response.contentUtf8());
+        assertThat(root.get("value").asText()).isEqualTo("a");
+        final JsonNode nested = root.get("nested");
+        assertThat(nested).isNotNull().matches(v -> ((TreeNode) v).isObject());
+        assertThat(nested.get("value").asText()).isEqualTo("b");
+    }
+
+    @Test
     void shoudDenyRecursiveViaHttpGet() throws Exception {
         assertThat(webClient.get("/v1/echo/recursive?value=a").aggregate().get().status())
+                .isEqualTo(HttpStatus.BAD_REQUEST);
+        assertThat(webClient.get("/v1/echo/recursive2?nested=a").aggregate().get().status())
                 .isEqualTo(HttpStatus.BAD_REQUEST);
 
         // Note that the parameter will be ignored because it won't be matched with any fields.

--- a/grpc/src/test/proto/com/linecorp/armeria/grpc/testing/transcoding.proto
+++ b/grpc/src/test/proto/com/linecorp/armeria/grpc/testing/transcoding.proto
@@ -112,6 +112,17 @@ service HttpJsonTranscodingTestService {
       }
     };
   }
+
+  rpc EchoRecursive(EchoRecursiveRequest) returns (EchoRecursiveResponse) {
+    option (google.api.http) = {
+      get: "/v1/echo/recursive"
+
+      additional_bindings {
+        post: "/v1/echo/recursive"
+        body: "value"
+      }
+    };
+  }
 }
 
 message GetMessageRequestV1 {
@@ -211,4 +222,17 @@ message EchoAnyRequest {
 
 message EchoAnyResponse {
   google.protobuf.Any value = 1;
+}
+
+message EchoRecursiveRequest {
+  Recursive value = 1;
+}
+
+message EchoRecursiveResponse {
+  Recursive value = 1;
+}
+
+message Recursive {
+  string value = 1;
+  Recursive nested = 2;
 }

--- a/grpc/src/test/proto/com/linecorp/armeria/grpc/testing/transcoding.proto
+++ b/grpc/src/test/proto/com/linecorp/armeria/grpc/testing/transcoding.proto
@@ -24,6 +24,8 @@ import "google/api/annotations.proto";
 import "google/protobuf/timestamp.proto";
 import "google/protobuf/duration.proto";
 import "google/protobuf/wrappers.proto";
+import "google/protobuf/struct.proto";
+import "google/protobuf/any.proto";
 
 service HttpJsonTranscodingTestService {
   rpc GetMessageV1(GetMessageRequestV1) returns (Message) {
@@ -64,6 +66,50 @@ service HttpJsonTranscodingTestService {
   rpc EchoWrappers(EchoWrappersRequest) returns (EchoWrappersResponse) {
     option (google.api.http) = {
       get: "/v1/echo/wrappers"
+    };
+  }
+
+  rpc EchoStruct(EchoStructRequest) returns (EchoStructResponse) {
+    option (google.api.http) = {
+      get: "/v1/echo/struct"
+
+      additional_bindings {
+        post: "/v1/echo/struct"
+        body: "value"
+      }
+    };
+  }
+
+  rpc EchoValue(EchoValueRequest) returns (EchoValueResponse) {
+    option (google.api.http) = {
+      get: "/v1/echo/value"
+
+      additional_bindings {
+        post: "/v1/echo/value"
+        body: "value"
+      }
+    };
+  }
+
+  rpc EchoListValue(EchoListValueRequest) returns (EchoListValueResponse) {
+    option (google.api.http) = {
+      get: "/v1/echo/list_value"
+
+      additional_bindings {
+        post: "/v1/echo/list_value"
+        body: "value"
+      }
+    };
+  }
+
+  rpc EchoAny(EchoAnyRequest) returns (EchoAnyResponse) {
+    option (google.api.http) = {
+      get: "/v1/echo/any"
+
+      additional_bindings {
+        post: "/v1/echo/any"
+        body: "value"
+      }
     };
   }
 }
@@ -133,4 +179,36 @@ message EchoWrappersResponse {
   google.protobuf.BoolValue boolVal = 7;
   google.protobuf.StringValue stringVal = 8;
   google.protobuf.BytesValue bytesVal = 9;
+}
+
+message EchoStructRequest {
+  google.protobuf.Struct value = 1;
+}
+
+message EchoStructResponse {
+  google.protobuf.Struct value = 1;
+}
+
+message EchoValueRequest {
+  google.protobuf.Value value = 1;
+}
+
+message EchoValueResponse {
+  google.protobuf.Value value = 1;
+}
+
+message EchoListValueRequest {
+  google.protobuf.ListValue value = 1;
+}
+
+message EchoListValueResponse {
+  google.protobuf.ListValue value = 1;
+}
+
+message EchoAnyRequest {
+  google.protobuf.Any value = 1;
+}
+
+message EchoAnyResponse {
+  google.protobuf.Any value = 1;
 }

--- a/grpc/src/test/proto/com/linecorp/armeria/grpc/testing/transcoding.proto
+++ b/grpc/src/test/proto/com/linecorp/armeria/grpc/testing/transcoding.proto
@@ -123,6 +123,18 @@ service HttpJsonTranscodingTestService {
       }
     };
   }
+
+  // When the request message is recursive.
+  rpc EchoRecursive2(Recursive) returns (Recursive) {
+    option (google.api.http) = {
+      get: "/v1/echo/recursive2"
+
+      additional_bindings {
+        post: "/v1/echo/recursive2"
+        body: "*"
+      }
+    };
+  }
 }
 
 message GetMessageRequestV1 {

